scripts/new_upstream_snapshot.py

```diff
@@ -0,0 +1,627 @@
+            elif line.startswith("Urgency"):
+                urgency = line.split(": ")[1]
+            elif line.startswith("Maintainer"):
+                maintainer = line.split(": ")[1]
+            elif line.startswith("Timestamp"):
+                timestamp = line.split(": ")[1]
+            elif line.startswith("Date"):
+                date = line.split(": ")[1]
+            elif line.startswith("Launchpad-Bugs-Fixed"):
+                bugs_fixed = line.split(": ")[1]
+            elif line.startswith("Changes"):
+                index = results.index("Changes:")
+                changes = results[index + len(line) + 1 :]
+                break
+            else:
+                raise RuntimeError(f"Could not parse line: {line}")
+        return cls(
+            source,
+            version,
+            distro,
+            urgency,
+            maintainer,
+            timestamp,
+            date,
+            bugs_fixed,
+            changes,
+        )
+
+
+def get_changelog_distro():
+    """Get the distro represented by this changelog.
+
+    The first line of d/changelog displays the distro. Since it can be
+    UNRELEASED, check the most recent entries until we find one.
+    """
+    for i in range(5):
+        details = ChangelogDetails.get(offset=i)
+        changelog_distro = details.distro
+        if changelog_distro != "UNRELEASED":
+            break
+    else:
+        raise RuntimeError("Could not determine distro from changelog")
+    return changelog_distro
+
+
+class EnvDetails(NamedTuple):
+    # This is a sloppy abstraction to avoid passing so much stuff around
+    changelog_details: ChangelogDetails
+    devel_distro: str
+    changelog_distro: str
+    head: str
+
+    @classmethod
+    def get(cls):
+        changelog_details = ChangelogDetails.get()
+        devel_distro = capture("distro-info --devel").stdout.strip()
+        changelog_distro = get_changelog_distro()
+        head = capture("git rev-parse HEAD").stdout.strip()
+        return cls(
+            changelog_details,
+            devel_distro,
+            changelog_distro,
+            head,
+        )
+
+
+def remove_line_from_file(filename, text):
+    """Utility function to remove a single line from a file."""
+    with open(filename, "r") as f:
+        lines = f.readlines()
+    with open(filename, "w") as f:
+        for line in lines:
+            if line.strip() != text.strip():
+                f.write(line)
+
+
+def merge_commitish(to: str) -> None:
+    """Perform 'git merge <commitish>' along with some error checking."""
+    try:
+        ref_name = capture(f"git describe --abbrev=8 {to}").stdout.strip()
+    except CalledProcessError as e:
+        raise RuntimeError(
+            f"'git describe' failed for {to}. "
+            "Is it a valid commtish or annotated tag?"
+        ) from e
+
+    command = f'git merge {to} -m "merge from {to} at {ref_name}"'
+    print(f"Running: {command}")
+    sh(command)
```

Comment:
This command will fail due to merge conflicts (fairly frequently thoughout the year). WDYT about making this a bit of an interactive section announcing the failure and letting us fix it instead of dropping out with a failure?
```suggestion
    try:
        capture(command)
    except CalledProcessError as e:
        conflict_paths = [
            l.split()[-1] for l in e.stdout.splitlines() if 'CONFLICT' in l
        ]
        if not conflict_paths:
            raise
        print(
            f"Found merge conficts: {' '.join(conflict_paths)}.\n"
            "Please follow these steps in a separate terminal:"
        )
        for conflict_path in conflict_paths:
           print(f"vi {conflict_path}; git add {conflict_path}")
        print("git merge --continue")
        if input("Did git merge --continue succced (y/N)? ").lower() != "y":
            raise
```

Reply:
`vi`? Get out of here :wink: 

If our dailies are building, we shouldn't ever have a merge conflict doing an upstream snapshot. But if we do still get conflicts, I would rather not have the script handle this. I prefer `git mergetool` to resolve conflicts, and doing something like this would just add extra friction. I would rather add a flag to continue after the merge is finished, or just make sure the branches are mergeable before you run the script.

Reply:
@TheRealFalcon that works for me. So, I suppose we need an exit point to handle this failure and a proper --short-circuit flag that can be passed to the script to jump to the specific failure stage we exited at. The reason I went with the former solution due to the fact that state isn't saved during an unexpected exception and I didn't want to add a 2nd flag --post-merge